### PR TITLE
Hand-specific Render Event

### PR DESCRIPTION
**Pull request by ChunHin**
#2747 Add hand-specific render event.
Example:
@SubscribeEvent
public void renderHand(RenderHandEvent event)
{
    if (event.getHand() == EnumHand.OFF_HAND)
    {
        event.setCanceled(true);
    }
}
It add which hand is rendering so that you can specific which hand not to render.

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch [hunk 1]
```diff
@@ -1,14 +1,79 @@
...
++import javax.imageio.ImageIO;
++import java.awt.*;
++import java.awt.image.BufferedImage;
++import java.awt.image.ImageObserver;
++import java.io.File;
++import java.io.IOException;
++import java.nio.FloatBuffer;
++import java.util.List;
++import java.util.Random;
++import java.util.concurrent.Callable;
```
Comment (RainWarrior) [inline patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch]: No import changes in patches

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch [hunk 2]
```diff
@@ -1,6 +1,57 @@
 --- ../src-base/minecraft/net/minecraft/client/renderer/ItemRenderer.java
 +++ ../src-work/minecraft/net/minecraft/client/renderer/ItemRenderer.java
-@@ -360,7 +360,7 @@
+@@ -302,13 +302,13 @@
+         GlStateManager.func_179109_b((float)i * 0.56F, -0.52F + p_187459_2_ * -0.6F, -0.72F);
+     }
+ 
+-    public void func_78440_a(float p_78440_1_)
```
Comment (RainWarrior) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: Don't remove any vanilla methods, especially public ones

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch [hunk 3]
```diff
@@ -1,6 +1,57 @@
...
++    public void renderItemInFirstPerson(float partialTicks, int pass)
```
Comment (RainWarrior) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: use forge.exc to make this method reobfuscate correctly

File: src/main/java/net/minecraftforge/client/event/RenderHandEvent.java [hunk 4]
```diff
@@ -1,20 +1,29 @@
...
+ */
 @Cancelable
 public class RenderHandEvent extends Event
 {
     private final RenderGlobal context;
     private final float partialTicks;
     private final int renderPass;
-    public RenderHandEvent(RenderGlobal context, float partialTicks, int renderPass)
+    private final EnumHand hand;
+    public RenderHandEvent(RenderGlobal context, float partialTicks, int renderPass, EnumHand hand)
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/client/event/RenderHandEvent.java]: Don't remove the old constructor, deprecate it. Binary compatibility.

File: patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch (hunk 1 above)
Comment (ChunHin) [inline patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch]: Fix it by changing the IDE settings

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch (hunk 2 above)
Comment (ChunHin) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: I recovered them

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch (hunk 3 above)
Comment (ChunHin) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: How to use it?

**Comment by ChunHin** [comment]
The second commit should fix the problems

File: patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch [hunk 5]
```diff
@@ -1,14 +1,22 @@
 --- ../src-base/minecraft/net/minecraft/client/renderer/EntityRenderer.java
 +++ ../src-work/minecraft/net/minecraft/client/renderer/EntityRenderer.java
-@@ -207,6 +207,7 @@
+@@ -43,7 +43,6 @@
+ import net.minecraft.entity.monster.EntitySpider;
+ import net.minecraft.entity.passive.EntityAnimal;
+ import net.minecraft.entity.player.EntityPlayer;
+-import net.minecraft.init.Blocks;
```
Comment (liach) [inline patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch]: No change on imports!

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch [hunk 6]
```diff
@@ -9,15 +28,15 @@
          {
              if (flag && this.field_187468_e == null)
              {
-@@ -444,6 +444,7 @@
+@@ -444,6 +449,7 @@
          if (this.field_78455_a.field_71439_g.func_70094_T())
          {
              IBlockState iblockstate = this.field_78455_a.field_71441_e.func_180495_p(new BlockPos(this.field_78455_a.field_71439_g));
 +            BlockPos overlayPos = new BlockPos(this.field_78455_a.field_71439_g);
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: what is this?

File: patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch (hunk 5 above)
Comment (williewillus) [inline patches/minecraft/net/minecraft/client/renderer/EntityRenderer.java.patch]: this whole patch seems unnecessary. Remove it.

File: patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch (hunk 6 above)
Comment (liach) [inline patches/minecraft/net/minecraft/client/renderer/ItemRenderer.java.patch]: The patch had it originally. It's not a new addition though.

**Comment by Illyohs** [comment]
This needs to be re-targeted to master

**Comment by ChunHin** [comment]
**EnumHand** only existed in 1.9

**Comment by covers1624** [comment]
1.9 has been merged in to master..

**Comment by ChunHin** [comment]
#2817 Re-targeted to master

**Pull request closed without merging**